tests/stub/optimizations/test_optimizations.py

```diff
@@ -0,0 +1,147 @@
+from nutkit.frontend import Driver
+import nutkit.protocol as types
+
+from tests.shared import (
+    driver_feature,
+    TestkitTestCase,
+)
+from tests.stub.shared import StubServer
+
+
+class TestOptimizations(TestkitTestCase):
+    def setUp(self):
+        super().setUp()
+        self._server = StubServer(9001)
+
+    def tearDown(self):
+        # If test raised an exception this will make sure that the stub server
+        # is killed and it's output is dumped for analysis.
+        self._server.reset()
+        super().tearDown()
+
+    @driver_feature(types.Feature.OPT_PULL_PIPELINING)
+    def test_pull_pipelining(self):
+        def test():
+            script = "pull_pipeline{}.script".format("_tx" if use_tx else "")
+            self._server.start(path=self.script_path(script),
+                               vars={"#TYPE#": mode[0]})
+            auth = types.AuthorizationToken(scheme="basic", principal="neo4j",
+                                            credentials="pass")
+            driver = Driver(self._backend, "bolt://%s" % self._server.address,
+                            auth)
+
+            session = driver.session(mode[0])
+            if use_tx:
+                tx = session.beginTransaction()
+                res = tx.run("CYPHER")
+                result = list(map(lambda r: r.value, res.next().values))
+                tx.commit()
+            else:
+                res = session.run("CYPHER")
+                result = list(map(lambda r: r.value, res.next().values))
+            self.assertEqual(result, [1])
+
+            session.close()
+            driver.close()
+            self._server.done()
+
+        for mode in ("read", "write"):
+            for use_tx in (True, False):
+                with self.subTest(mode + ("_tx" if use_tx else "")):
+                    test()
+                self._server.reset()
+
+    def double_read(self, mode, new_session, use_tx, check_single_connection,
```

Comment:
This method could be moved to the end of the file to keep the tests methods on the top. It will make it easier to navigate through the file.

Reply:
Most other test files have the common code above the test methods. I don't have a strong opinion about this, so I'll go with consistency.